### PR TITLE
upgrade hackport to actual cabal

### PR DESCRIPTION
**Pull request by Miezhiko**
WIP
===

This is dirty draft but I may use review and I need help with porting update method

### CODE REVIEW

**Comment by hololeap** [comment]
Yes, we got stuck with the update function as well. May I ask, what is "actual cabal" in this context?

**Comment by hololeap** [comment]
I'd like to get it so that hackport just depends on the Cabal library instead of having it as a submodule, if that's possible. Then `hackport update` could just run `cabal update` or tell the user to do so.

File: src/Hackport/Command/Update.hs [hunk 1]
```diff
@@ -1,16 +1,17 @@
 module Hackport.Command.Update
-  ( updateAction
+  ( CabalInstall.updateAction
   ) where
 
 import Distribution.Simple.Command (CommandUI(..))
```
Comment (hololeap) [inline src/Hackport/Command/Update.hs:5]: The CI runs with the `pedantic` flag enabled, which turns on `-Werror`. So, you'll have to fix any warnings for the tests to pass.

**Comment by Miezhiko** [comment]
> I'd like to get it so that hackport just depends on the Cabal library instead of having it as a submodule, if that's possible. Then `hackport update` could just run `cabal update` or tell the user to do so.

agree, that could be better, I wasn't sure if gentoo-haskell fork has something very specific

File: src/Hackport/Command/Update.hs (hunk 1 above)
Comment (Miezhiko) [reply src/Hackport/Command/Update.hs:5]: > The CI runs with the `pedantic` flag enabled, which turns on `-Werror`. So, you'll have to fix any warnings for the tests to pass.

at first I need to do something with update command, maybe running CmdUpdate from cabal-install is good way but need to rework arguments used, new cabal install wants some "NixStyle opts" to be passed.

**Comment by Miezhiko** [comment]
> Yes, we got stuck with the update function as well.
> 
> > upgrade hackport to actual cabal
> 
> What is "actual cabal" in this context?

old cabal fork seems like can't even be built with ghc 9 2 4 and newer, so I've just used  to rebase the fork on upstream master

**Comment by hololeap** [comment]
It looks like the `3.8` line of `Cabal` exposes a lot more API (especially `cabal-install` which used to expose none), so once we get that into `gentoo-haskell`, we could work on getting the submodules out of this repo and just depending on libs normally. I did some experiments last night and found that I could get it to compile without the submodules with [surprisingly few tweaks](https://github.com/hololeap/hackport/tree/remove-bundled-libs), but unfortunately I got `Assertion` exceptions from deep inside `Cabal-3.8` on some tests and when I tried to run `hackport merge`.

> I need help with porting update method

I want to prioritize getting this PR pushed through first. I have a bit more familiarity with the `Cabal` codespace after tinkering last night, and I think I might be able to get the `hackport update` command working with the updated submodules.

> I wasn't sure if gentoo-haskell fork has something very specific

@solpeth has done a lot more work on this project than I have and probably knows why we have been using forks for `Cabal`/`cabal-install`/`hackage-security`.

**Comment by Miezhiko** [comment]
@hololeap I'd go with GHC upgrade first (from other PR), but not fully sure.

**Comment by hololeap** [comment]
> @hololeap I'd go with GHC upgrade first (from other PR), but not fully sure.

It's going to be a while before we keyword `ghc-9.4`, if that's what you're suggesting. It's quite a bit easier to get `cabal*` bumped up from what I've seen. In any case, I just wanted to mention that I had some news in regard to removing the submodules, but I think getting the `cabal` submodule up-to-date, like you suggested here, is a much better first step.

**Comment by hololeap** [comment]
Here are the custom changes that exist in the `gentoo-haskell/cabal` repo:

- https://github.com/hololeap/cabal/commit/eb60cc9533b527101cff62e6e422706ff5704a04
- https://github.com/hololeap/cabal/commit/b83d3b9121448984a9949ed25605ffa04270400c
- https://github.com/hololeap/cabal/commit/89f89c99fea31eefd809955a8d281acacceeb375
- https://github.com/hololeap/cabal/commit/0f27623abbc6c3455f0f6009ede5cda248a7c0d1
- https://github.com/hololeap/cabal/commit/3f028dea3727e9ad71d75ec51a7f6b66c71b816a
- https://github.com/hololeap/cabal/commit/3615d7f1da6b79e68e3768c5d912145187780912
- https://github.com/hololeap/cabal/commit/849b1878f19f9291b813e75a1ae4b796144b1a06
- https://github.com/hololeap/cabal/commit/5a9ccd28702efa754aa9f996c886fbf820900b3a
- https://github.com/hololeap/cabal/commit/5b3ffd7de3497ee712650176b35d2c0d11181b51
- https://github.com/hololeap/cabal/commit/3da957aa1e44568bc3b39d4c8c9f49a6c838dc0d
- https://github.com/hololeap/cabal/commit/aeb5960d0ebeb37f5aeb3fae707cbcdcca8d3974
- https://github.com/hololeap/cabal/commit/e055adec4c46f8455a118834130480fd29f97f43
- https://github.com/hololeap/cabal/commit/a3b85656e92583779616420182b987e7939a4513

I managed to get those changes rebased on the latest `master` branch from `haskell/cabal`, so I will try switching my copy of your branch to this rebased `cabal` submodule and see how things go.

**Comment by hololeap** [comment]
@Miezhiko I have a question on your commit message

> TODO: fix update (who is using it anyways?)

Do you _not_ run `hackport update`? If not, how are you getting up-to-date package lists from hackage?

**Comment by Miezhiko** [comment]
> Here are the custom changes that exist in the `gentoo-haskell/cabal` repo:
> 
> * [hololeap/cabal@eb60cc9](https://github.com/hololeap/cabal/commit/eb60cc9533b527101cff62e6e422706ff5704a04)
> * [hololeap/cabal@b83d3b9](https://github.com/hololeap/cabal/commit/b83d3b9121448984a9949ed25605ffa04270400c)
> * [hololeap/cabal@89f89c9](https://github.com/hololeap/cabal/commit/89f89c99fea31eefd809955a8d281acacceeb375)
> * [hololeap/cabal@0f27623](https://github.com/hololeap/cabal/commit/0f27623abbc6c3455f0f6009ede5cda248a7c0d1)
> * [hololeap/cabal@3f028de](https://github.com/hololeap/cabal/commit/3f028dea3727e9ad71d75ec51a7f6b66c71b816a)
> * [hololeap/cabal@3615d7f](https://github.com/hololeap/cabal/commit/3615d7f1da6b79e68e3768c5d912145187780912)
> * [hololeap/cabal@849b187](https://github.com/hololeap/cabal/commit/849b1878f19f9291b813e75a1ae4b796144b1a06)
> * [hololeap/cabal@5a9ccd2](https://github.com/hololeap/cabal/commit/5a9ccd28702efa754aa9f996c886fbf820900b3a)
> * [hololeap/cabal@5b3ffd7](https://github.com/hololeap/cabal/commit/5b3ffd7de3497ee712650176b35d2c0d11181b51)
> * [hololeap/cabal@3da957a](https://github.com/hololeap/cabal/commit/3da957aa1e44568bc3b39d4c8c9f49a6c838dc0d)
> * [hololeap/cabal@aeb5960](https://github.com/hololeap/cabal/commit/aeb5960d0ebeb37f5aeb3fae707cbcdcca8d3974)
> * [hololeap/cabal@e055ade](https://github.com/hololeap/cabal/commit/e055adec4c46f8455a118834130480fd29f97f43)
> * [hololeap/cabal@a3b8565](https://github.com/hololeap/cabal/commit/a3b85656e92583779616420182b987e7939a4513)
> 
> I managed to get those changes rebased on the latest `master` branch from `haskell/cabal`, so I will try switching my copy of your branch to this rebased `cabal` submodule and see how things go.
> 
> Edit:
> 
> It looks like `master` may be too far ahead, so I'll try compiling/testing after rebasing on the `3.8` branch.

yes, I've rebased those too, changes in this pull request are targeting my cabal fork of gentoo-haskell's cabal

> Do you not run hackport update? If not, how are you getting up-to-date package lists from hackage?

> (I'm just curious if you've found another way to do it that is working for you)

I was thinking about something else when I was writing this comment, but yes I've found another way, I just copy needed files after cabal update to `.hackport` directory, this even possible to automate with some script.

**Comment by hololeap** [comment]
> I was thinking about something else when I was writing this comment, but yes I've found another way, I just copy needed files after cabal update to .hackport directory, this even possible to automate with some script.

Good to know the updated submodules make Hackport compatible with the Cabal directory. That might mean replacing `hackport update` with `cabal update` will be an easy transition.

**Comment by hololeap** [comment]
Please test [my fork](https://github.com/hololeap/hackport) and see if it works for you on ghc-9.2 and what it takes to get the package directory working on your system. I think there is an incompatibility with how the last version of Hackport stored its directory compared to the new one -- I had to delete `.hackport` and run `hackport update` -- but it might work with your method of copying from `~/.cabal`.

**Comment by Miezhiko** [comment]
> Please test [my fork](https://github.com/hololeap/hackport) and see if it works for you on ghc-9.2 and what it takes to get the package directory working on your system. I think there is an incompatibility with how the last version of Hackport stored its directory compared to the new one -- I had to delete `.hackport` and run `hackport update` -- but it might work with your method of copying from `~/.cabal`.

can't test in near time, minimum in middle of next week or even much later

**Comment by hololeap** [comment]
> can't test in near time, minimum in middle of next week or even much later

No worries! I'll get this pushed into `gentoo-haskell/hackport` soon and probably release it as a `**` keyworded ebuild, so feel free to test it when you can. Thanks for your contribution!

**Comment by Miezhiko** [comment]
@hololeap Configuring hackport-0.8.0.0...

setup: Encountered missing or private dependencies:
process >=1.6.15.0

**Comment by Miezhiko** [comment]
@hololeap tested, works good!

**Comment by hololeap** [comment]
@Miezhiko Awesome! Glad to hear it.

**Pull request closed without merging**